src/docs/arch-compact.dox

```diff
@@ -0,0 +1,57 @@
+starts.  The compaction logic then starts examining the files that need to be compacted. Compacting
+the object is done 10% at a time, that is, compaction tries to move blocks from the last 10% of the
+file into the beginning of the file (the 10% is hard coded in the block manager).  The reason for
+this is because compaction walks the file in logical order, not block offset order, and compaction
+of a file can fail if the block from the end of the file is not written first. Note that the block manager
+uses a first-fit block selection algorithm during compaction to maximize block movement. The process
+is repeated multiple times until block manager detects that there no blocks from the last 10% of the file
+that can be moved.
+
+Diving deep into the compaction process, the btree layer walks the internal pages for a given table to
+identify leaf pages that can be rewritten. For each leaf page referenced by the internal page, we use
+a block manager API to assess whether it will be beneficial to rewrite the leaf page. If the block
+manager recognizes that rewriting the page will be beneficial, it rewrites the disk
+block to a new disk location. The new disk address is placed into the parent internal page and
+the page is marked dirty. In this way the leaf page is rewritten without being brought into the
+cache. If the leaf page is already in cache, the leaf page is simply marked dirty and the subsequent
+checkpoint or eviction rewrites the page.
+
+After each 10% compaction, checkpoint is performed two more times. The second and third
+checkpoints are because the block manager checkpoints in two steps: blocks made available for reuse
+during a checkpoint are put on a special checkpoint-available list and only moved to the real available
+list after the metadata has been updated with the new checkpoint's information. For this reason,
+blocks allocated to write the checkpoint itself cannot be taken from the blocks made available by
+the checkpoint. To say it another way, the second checkpoint puts the blocks from the end of the file
+that were made available by compaction onto the checkpoint-available list, but then potentially
+writes the checkpoint itself at the end of the file, which would prevent any file truncation.  When
+the metadata is updated for the second checkpoint, the blocks freed by compaction become available for
+the third checkpoint, so the third checkpoint's blocks are written towards the beginning of the file,
+and then the file can be truncated.
+
+Compaction is a time consuming operation and therefore logs have been added in compaction logic to
+track current progress. Some of the verbose logs for compaction can be activated by setting
+\c "verbose=[compact,compact_progress]" configuration option for \c wiredtiger_open() method. Apart from
+verbose logs, WT also records statistics about compact progress that can be used to debug issues
+with compaction process.
```

Comment:
Maybe we should mention `__VERBOSE_OUTPUT_PERCENTILE` definition that provides even more detailed verbose output?

Reply:
I would say `__VERBOSE_OUTPUT_PERCENTILE` is something that end user can not set through an API and therefore I would avoid mentioning this here. This doc is also intended for readers that do not deal with WT codebase.